contribute/design/1.x/cstor/20200511-cstor-backupandrestore.md

```diff
@@ -0,0 +1,323 @@
+      - [Work Flow](#work-flow)
+        - [Velero Server sends CreateSnapshot Request](#velero-server-sends-createsnapshot-request)
+        - [Velero Server Sends DeleteSnapshot Request](#velero-server-sends-deletesnapshot-request)
+        - [Velero Server Sends CreateVolumeFromSnapshot Request](#velero-server-sends-createvolumefromsnapshot-request)
+        - [Accessing CVC REST Endpoint](#accessing-cvc-rest-endpoint)
+        - [Velero Server Sends DeleteSnapshot Request](#velero-server-sends-deletesnapshot-request-1)
+        - [Work flow in backup controller](#work-flow-in-backup-controller)
+        - [Work flow in restore controller](#work-flow-in-restore-controller)
+  - [Design Details](#design-details)
+    - [CStorBackup Schema](#cstorbackup-schema)
+    - [CStorRestore Schema](#cstorrestore-schema)
+    - [CStorCompletedBackup Schema](#cstorcompletedbackup-schema)
+  - [Drawbacks](#drawbacks)
+  - [Alternatives](#alternatives)
+  - [Infrastructure Needed](#infrastructure-needed)
+
+## Summary
+
+This proposal brings out the design details to implement backup and restore
+solution for V1 version of CStorVolumes.
+
+## Motivation
+
+- Create a backup of the cStor persistent volume to the desired storage location(
+  this can be AWS, GCP or any storage provider). This backup will be either on-demand or scheduled backup.
+- Restore this backup to the same cluster or another cluster.
+
+### Goals
+
+- Solution to create an on-demand backup of cStor persistent volumes.
+- Solution to create an on-demand restore of backuped up cStor persistent volumes.
+- Solution to create a scheduled backup of cStor persistent volumes.
+- Solution to create an incremental/differential backup of cStor persistent volumes. With incremental/differential backup, the user will be able to save backup storage space and it 
+
+### Non-Goals
+
+- Supporting local backup and restore.
+
+## Proposal
+
+### User Stories
+
+#### Create a backup for CStorVolumes
+As an OpenEBS user, I should be able create a backup for CStorVolumes.
+
+#### Create a restore for backuped CStorVolumes
+As an OpenEBS user, I should be restore a backuped CStorVolumes.
+
+#### Scheduled backup of CStorVolumes
+As an OpenEBS user, I should be able to create a scheduled backups for CStorVolumes.
+
+### Proposed Implementation
+
+#### CVC-Operator REST Interface
+
+The volume management api like create, delete, snapshot, clone etc have moved from m-apiserver to CVC-Operator as part of supporting the CSI Driver. The velero-plugin of volume API also should be moved to the CVC Operator as the velero-plugin api in turn depends on the volume snapshot, clone implementations. Currently CVC Operator only supports declarative API via CRs, but plugin requires imperative API(REST API). The proposal is to implement a REST server within CVC-Operator to perform velero-plugin operations. The velero-plugin should identify the type of the volume and forward the REST API requests to the CVC Operator.
```

Comment:
`The velero-plugin of volume API also should be moved to the CVC Operator as the velero-plugin api in turn depends on the volume snapshot, clone implementations.` Can we put it this way, "REST API of Maya-ApiServer to trigger backup/restore also needs to be implemented in the CVC operator as these APIs are using volume snapshot/clone API."

Reply:
updated